### DISCUSSION TITLE
Substream not modulo 16

### DISCUSSION THREAD

**Discussion by Fluffkin** [Motion-Project/motion#1515] [General]
Yes checked docs.
Version 4.4.0
Installed from package as standalone
Video source netcam rtsp
Running on Pi3B+
Void OS (32bit arm)

I believe modulo 8 has been implemented for quite some time, so I'm somewhat surprised to get "Substream not available.  Image sizes not modulo 16" errors in the log files from line 1158 in motion.c

I have a netcam_url for the low rez motion detection of 640x360, and a netcam_high_url at 1920x1080 set to pass through mode.
I'm setting the width / height as 640x360 to avoid rescaling. Without that setting the preview images in motion's html frontend are the wrong ratio. Presumably rescaled to the default width/height.

As far as I can tell it all still works fine. I'm just puzzled why the error appears (and if it's causing issues I don't know about).

**Comment by Mr-Dave** [comment]
It just means the a web stream substream is not available.  If you needed a substream, then it would be a critical issue....if you don't need a substream, it isn't something to worry about.

**Reply by Fluffkin** [reply]
Thanks for the reply. But in camera terminology, the "substream" is the low resolution rtsp feed. I'm trying to use that for motion detection.

You're saying that the substream is not available. Is that saying that the camera low resolution feed is not available to Motion? Or that Motion uses substream to mean something else? If it's the first, then I presume Motion is using the hi-res feed for motion detection? Or how is it managing to detect motion if the low-rez stream is not available?

**Reply by Mr-Dave** [reply]
You are mixing up inputs vs outputs.  Motion creates its own output streams.  Therefore, the substream from Motion is the low-resolution output version regardless of the input it is getting from the camera.  The "stream" is full resolution of whatever is being provided.  

https://motion-project.github.io/motion_config.html#OptDetail_Stream

**Reply by Fluffkin** [reply]
Ahh, I hadn't considered the streams. The only time I see them is in the html frontend.

Perhaps changing the error message to "Motion generated output Substream not available..." would be more helpful? It probably would have given me the clue I needed to find that part of the documentation, and I would have known not to worry that it meant the camera stream.

Thanks for the explanation.

**Comment by BugerDread** [comment]
Hallo, Im trying to set up motion with FHD cameras (1920x1080 main and 640x360 lo-res rtsp stream).

I found that substreams (like <ip-of-motion-server>:<port>/substream) are just the same as main streams eg 1920x1080 (but should be half in every dimension). In logs I found the same message mentioned here: "Substream not available. Image sizes not modulo 16". 

I managed to get substream working by a "hacky" way by setting the resolution to 1920x1072 instead of 1080 to match "modulo 16" using
```
width 1920
height 1072
```
but the CPU cost of the rescaling to 1072 is quite high - as mentioned in doc

Is there any other way how to get substreams working?

Unfortunately 1080 is not modulo 16, 360 is also not modulo 16, so it is not possible to save bw using substreams with FHD cams?

**Reply by Fluffkin** [reply]
As Mr-Dave explained above. The modulo 16 error message concerns the webstream produced by Motion, an MJPEG feed you can view produced by Motion on the network. If you don't need it (I don't). Then you don't need to worry about it. If you really want to watch the live stream from the camera it's usually easier to point mpv / vlc etc directly at the webcam feed and not at motion's produced MJPEG stream.
